### PR TITLE
Cal 204 free slots in chunks

### PR DESCRIPTION
**Pull request by qxf2**

### CODE REVIEW

File: qxf2_scheduler/qxf2_scheduler.py [hunk 1]
```diff
@@ -4,12 +4,24 @@
 """
 
 import qxf2_scheduler.base_gcal as gcal
+#import base_gcal as gcal
```
Comment (qxf2) [inline qxf2_scheduler/qxf2_scheduler.py]: This line should be deleted.

File: qxf2_scheduler/qxf2_scheduler.py [hunk 2]
```diff
@@ -38,7 +50,83 @@ def is_weekend(date):
...
+            #Initializing the free slot end            
+            free_slot_start = free_slots[i]['start']
+            #Intializing the next free slot start        
+            free_slot_end = free_slots[i]['end']
+            
+            #Find the difference between start and end slot
+            diff_between_slots = convert_string_into_time(free_slot_end) - convert_string_into_time(free_slot_start)
+            
+            if diff_between_slots >= timedelta(minutes=30):                
+                if free_slot_start[-2:]=='00':
```
Comment (qxf2) [inline qxf2_scheduler/qxf2_scheduler.py]: Can you make lines 73-93 into two methods. That way, we can simply do:

free_slot_start = get_modified_free_slot_start(free_slots[i]['start'],marker='30')
free_slot_end = get_modified_free_slot_end(free_slots[i]['end'],marker='30')

File: qxf2_scheduler/qxf2_scheduler.py [hunk 3]
```diff
@@ -38,7 +50,83 @@ def is_weekend(date):
...
+
+
+def get_free_slots_in_chunks(free_slots):
+    "Return the free slots in 30 minutes interval"
+    #Appending the 30 minutes slot into list
+    if free_slots == None:
+        print("There are no more free slots available for this user")    
+    else:        
+        chunk_time_interval = []        
+        for i in  range(0,len(free_slots)):
```
Comment (qxf2) [inline qxf2_scheduler/qxf2_scheduler.py]: Why this style of for loop? Couldn't you just do:

for free_slot in free_slots:

File: qxf2_scheduler/qxf2_scheduler.py [hunk 4]
```diff
@@ -38,7 +50,83 @@ def is_weekend(date):
...
+            #Initializing the free slot end            
+            free_slot_start = free_slots[i]['start']
+            #Intializing the next free slot start        
+            free_slot_end = free_slots[i]['end']
+            
+            #Find the difference between start and end slot
+            diff_between_slots = convert_string_into_time(free_slot_end) - convert_string_into_time(free_slot_start)
+            
+            if diff_between_slots >= timedelta(minutes=30):
```
Comment (qxf2) [inline qxf2_scheduler/qxf2_scheduler.py]: Can you make sure not to hard code 30 everywhere? Instead, pass it to get_free_slots_in_chunks as an argument with default value 30. 

We will need to be changing this depending on the interview round, no? So in the coming iterations, we can change the logic to the way we calculate slot start, slot end and appending without having to change numbers.

File: qxf2_scheduler/qxf2_scheduler.py [hunk 5]
```diff
@@ -38,7 +50,83 @@ def is_weekend(date):
...
+                if free_slot_end[-2:]=='00' or free_slot_end[-2:]=='30' :
+                    modified_free_slot_end = free_slot_end                    
+
+                elif free_slot_end[-2:] <'30':                    
+                    modified_free_slot_end = '{}:{}'.format(free_slot_end.split(':')[0], '00')
+                    
+                elif free_slot_end[-2:] > '30':
+                    modified_free_slot_end = '{}:{}'.format(free_slot_end.split(':')[0], '30')                    
+                
+                result_flag = True
```
Comment (qxf2) [inline qxf2_scheduler/qxf2_scheduler.py]: Lines 95-128 should become a method of its own. Something like `get_chunks_in_slot(slot_start,slot_end, chunk_size):`

That way your `get_free_slots_in_chunks(free_slots)` method looks something like this:

```
def get_free_slots_in_chunks(free_slots,chunk_size=30):
  "Return chunked free slots"
  for free_slot in free_slots:
    chunk_time_intervals = []
    free_slot_start = get_modified_free_slot_start(free_slot['start'],marker=chunk_size)
    free_slot_end = get_modified_free_slot_end(free_slot['end'],marker=chunk_size)
    slot_duration = convert_string_into_time(free_slot_end) - convert_string_into_time(free_slot_start)
    if slot_duration >= chunk_size:
      chunks = get_chunks_in_slot(free_slot_start,free_slot_end, chunk_size=chunk_size)
      chunk_time_intervals += chunks
  
   return chunk_time_intervals
```

File: qxf2_scheduler/routes.py [hunk 6]
```diff
@@ -5,6 +5,7 @@
 from flask import render_template, url_for, flash, redirect, jsonify, request, Response
 from qxf2_scheduler import app
 import qxf2_scheduler.qxf2_scheduler as my_scheduler
+import sys
```
Comment (qxf2) [inline qxf2_scheduler/routes.py]: Why do we need this import?

File: qxf2_scheduler/templates/get-schedule.html [hunk 7]
```diff
@@ -113,22 +113,28 @@
                         $("#resultDiv").css('color', 'black');
                         console.log(result);
                         freeSlots = result.free_slots;
```
Comment (qxf2) [inline qxf2_scheduler/templates/get-schedule.html]: Do we still need this line?

**Comment by qxf2** [comment]
@Annapooraniqxf2 - overall looks good. I have given my feedback as comments on each file. The big picture feedback is that you need to:

a) Break the method you wrote into multiple smaller pieces (something like [this](https://github.com/qxf2/interview-scheduler/pull/1#discussion_r309844538))
b) You should try to get rid of the hard code for 30 minutes.

File: qxf2_scheduler/qxf2_scheduler.py (hunk 1 above)
Comment (Annapooraniqxf2) [reply qxf2_scheduler/qxf2_scheduler.py]: Done

File: qxf2_scheduler/qxf2_scheduler.py (hunk 2 above)
Comment (Annapooraniqxf2) [reply qxf2_scheduler/qxf2_scheduler.py]: Done

File: qxf2_scheduler/qxf2_scheduler.py (hunk 3 above)
Comment (Annapooraniqxf2) [reply qxf2_scheduler/qxf2_scheduler.py]: Done

File: qxf2_scheduler/qxf2_scheduler.py (hunk 4 above)
Comment (Annapooraniqxf2) [reply qxf2_scheduler/qxf2_scheduler.py]: Done

File: qxf2_scheduler/qxf2_scheduler.py (hunk 5 above)
Comment (Annapooraniqxf2) [reply qxf2_scheduler/qxf2_scheduler.py]: done

File: qxf2_scheduler/routes.py (hunk 6 above)
Comment (Annapooraniqxf2) [reply qxf2_scheduler/routes.py]: Done

File: qxf2_scheduler/templates/get-schedule.html (hunk 7 above)
Comment (Annapooraniqxf2) [reply qxf2_scheduler/templates/get-schedule.html]: Done

**Pull request closed without merging**